[UI Framework] Improve KuiContextMenu keyboard navigation UX

cjcenizal opened the pull request:
Resolves https://github.com/elastic/kibana/issues/14239
Also addresses https://github.com/elastic/kibana/pull/14183#pullrequestreview-68050231

@timroes I implemented your suggestions for the improved keyboard navigation. @weltenwort I refactored the code a bit and got rid of that `isVisible` prop. It turns out it wasn't necessary at all --embarrassingly, it didn't even do what I originally thought it was doing.

tylersmalley commented:
jenkins, test it

cjcenizal commented:
jenkins, test this

timroes commented:
@cjcenizal you have to merge master into this, to get rid of the licencing issue. I merged a fix for that yesterday.

timroes commented:
I think this PR destroyed (on purpose?) some of the old behavior.

In panels that are menu only, I cannot use <kbd>Shift</kbd>+<kbd>Tab</kbd> anymore to move focus to the header (and press enter there to move an layer up). I can now only use <kbd>Left</kbd> to move a layer up. But the content dialog in the example (i.e. a non menu dialog?) can still focus that header and use enter.

Is this a desired behavior or got this broken by accident?

timroes commented:
Due to the `overflow: hidden` on the `.kuiContextMenu`the focus highlight isn't shown in the first example in the docs, but in the third it's shown.

To be honest I don't find that focus ring in this place very attractive and I would like to discuss other solutions.

I would either try to create a more beautiful focus ring for this, though I had a hard time coming up with an idea how this could look like, or get rid of the focus ring for the dialog in general.

*Why would I even say such a thing?* Here some arguments why I think we can get rid of it in this place:

* Having focus rings is WCAG AA - so even they don't consider it the very most important thing to place it in A level compliance
* Even though the element is focused, you cannot interact with it via keyboard in this moment. So pressing Enter or Space doesn't do anything, so I think it's not as important to know, that this element currently has the focus.
* The popover is seen as kind of a modal over the page. If you click outside it, it vanished, if you press Escape, it vanishes. So as soon as it opens the attention of the user is anyway at it. Also pressing Tab in that moment will directly interact with the dialog, same as the arrow keys.

I just don't see a good reason, that this focus ring would add much for the user in this case. That she would expect more likely that pressing keys will now interact with it? Wouldn't most users assume that anyway for something behaving very modal like? I just feel we lose too much from a design perspective for gaining little to nothing in usability.

cjcenizal commented:
@timroes Thanks for those terrific catches! I completely agree that the focus ring provides no benefit and sends the wrong message to the user (I missed that originally). I've removed it as you suggested. I also fixed that bug with the tabbing UX -- that was a regression. Could you take another look?

weltenwort commented on ui_framework/src/components/context_menu/context_menu_panel.js:
```diff
@@ -220,7 +234,13 @@ export class KuiContextMenuPanel extends Component {
+    if (node) {
+      this.content = node;
+    }
+  };
```
If we store a reference to the dom element here, we should also clear it when we're called with `null`. Just storing `this.content = node` would do that.

cjcenizal commented:
@weltenwort Bah thanks for catching that mistake in the tests. I've addressed both of your points. Mind taking one last quick look?

weltenwort commented on ui_framework/src/components/context_menu/context_menu_panel.js:
```diff
@@ -226,20 +226,19 @@ export class KuiContextMenuPanel extends Component {
+    this.panel = node;
+
+    if (this.panel) {
       this.panel.addEventListener('animationend', this.onTransitionComplete);
```
Now if we just could remove the event listener again when the previous `this.panel` was an element but the `node` isn't... or even better replace it with the `onAnimationEnd` react prop :thought_balloon:

cjcenizal commented:
@weltenwort Done and done! You are my new favorite code reviewer.

timroes commented:
Okay now I unfortunately must start nitpicking a bit 😅 

When pressing <kbd>Tab</kbd> you can cycle through all options, and it will start from the top (i.e. the dialog itself) again when reaching the end. The same behavior doesn't work when using <kbd>Shift</kbd>+<kbd>Tab</kbd> to cycle backwards. You will be stuck on the dialog in this case.

If this is an actual issue in one of the libraries, I wouldn't consider it blocking or too worse, just crossed my mind.

timroes commented:
As discussed in person: If the trigger button is already focused, and you scroll away, pressing <kbd>Enter</kbd> will open the menu, but not scroll back to the menu (in contrast to all other actions with the menu). Very nitpicking, but you might want to look into it :-)

weltenwort commented on ui_framework/src/components/context_menu/context_menu_panel.js:
```diff
@@ -118,53 +156,29 @@ export class KuiContextMenuPanel extends Component {
+    // If there aren't any items then this is probably a form or something.
+    // So let's focus the first tabbable item and expedite input from the user.
+    if (!this.menuItems.length) {
+      const tabbableItems = tabbable(this.content);
```
Maybe we should guard against `this.content` being `null | undefined`?

cjcenizal commented:
@timroes @weltenwort Comments addressed! Could you take another look?

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins, test this
